Add some basic "Getting Started" documentation for the AI packages (#1098)
Effect-TS/website: content/src/content/docs/docs/ai/getting-started.mdx

npm install @effect/ai-openai

# Also add the core Effect package (if not already installed)
npm install effect
```

</TabItem>

<TabItem label="pnpm" icon="pnpm">

```sh showLineNumbers=false
# Install the base package for the core abstractions (always required)
pnpm add @effect/ai

# Install one (or more) provider integrations
pnpm add @effect/ai-openai

# Also add the core Effect package (if not already installed)
pnpm add effect
```

</TabItem>

<TabItem label="Yarn" icon="seti:yarn">

```sh showLineNumbers=false
# Install the base package for the core abstractions (always required)
yarn add @effect/ai

# Install one (or more) provider integrations
yarn add @effect/ai-openai

# Also add the core Effect package (if not already installed)
yarn add effect
```

</TabItem>

<TabItem label="Bun" icon="bun">

```sh showLineNumbers=false
# Install the base package for the core abstractions (always required)
bun add @effect/ai

# Install one (or more) provider integrations
bun add @effect/ai-openai

# Also add the core Effect package (if not already installed)
bun add effect
```

</TabItem>

</Tabs>

## Define an LLM Interaction

First let's define a simple interaction with an LLM:

> ```ts twoslash

Reviewer: I'll add a title for the example here, for example

```md
**Example** (Using the Completions service to generate a dad joke)
```

import { Completions } from "@effect/ai"
import { Console, Effect } from "effect"

// Using `Completions` will add it to your program's requirements
//
//          ┌─── Effect<void, AiError, Completions>
//          ▼
const generateDadJoke = Effect.gen(function*() {
  // Extract the `Completions` service from the Effect environment
  const completions = yield* Completions.Completions
  // Use the `Completions` service to generate some text

>   return yield* completions.create("Generate a hilarious dad joke").pipe(
>     // Log the response text to the console
>     Effect.andThen((response) => Console.log(response.text))
>   )

Reviewer: ```suggestion
  // Use the `Completions` service to generate some text
  const res = yield* completions.create("Generate a hilarious dad joke")
  // Log the response text to the console
  console.log(res.text)
  return res
```

Author: In this case I don't want to return the response, I just want to log out the generated text.

Reviewer: The reason I suggest returning a value instead of `void` is because I found it confusing later on.

In the following snippets, I was wondering where my dad jokes went and how I could access or manage them:

```ts
const main = Effect.gen(function*() {
  // Build the `AiModel` into a `Provider`
  const gpt4o = yield* Gpt4o
  // Provide the implementation of `Completions` to `generateDadJoke`
  yield* gpt4o.provide(generateDadJoke)
})
```

```ts
const main = Effect.gen(function*() {
  const gpt = yield* Gpt4o
  yield* gpt.provide(generateDadJoke)
  yield* gpt.provide(generateDadJoke)
  yield* gpt.provide(generateDadJoke)
})
```

I think returning the response (which is ultimately what I'm trying to get from an LLM request) would be clearer and more intuitive than returning `void`.

})
```

> The above program uses the `Completions` service from the base `@effect/ai` package to describe using an LLM to create a text completion (in this case generating a hilarious dad joke). The result of the text generation is then logged to the console.

Reviewer: this seems a bit redundant

<Aside type="note" title="Declarative LLM Interactions">
  Notice that the above code does not know or care which LLM provider (OpenAi, Anthropic, etc.) will be used. Instead, we focus on _what_ we want to accomplish (i.e. our business logic), not _how_ to accomplish it.
</Aside>

## Select an LLM Model

Next, we need to select which provider we want to use to satisfy the `Completions` requirement of our `generateDadJoke` program:

> 

Reviewer: ```md
**Example** (Providing a concrete model to satisfy the `Completions` requirement)
```

> ```ts twoslash {1,17,25,27} collapse={5-10}

Reviewer: Too many lines are highlighted, I think the code is clear enough without them

import { OpenAiCompletions } from "@effect/ai-openai"
import { Completions } from "@effect/ai"
import { Console, Effect } from "effect"

const generateDadJoke = Effect.gen(function*() {
  const completions = yield* Completions.Completions
  return yield* completions.create("Generate a hilarious dad joke").pipe(
    Effect.andThen((response) => Console.log(response.text))
  )
})

// Create an `AiModel` which provides a concrete implementation of 
// `Completions` and requires an `OpenAiClient`
//
//      ┌─── AiModel<Completions, OpenAiClient>
//      ▼
const Gpt4o = OpenAiCompletions.model("gpt-4o")

// Building an `AiModel` adds its requirements to the program
//
//     ┌─── Effect<void, AiError, OpenAiClient>
//     ▼
const main = Effect.gen(function*() {
  // Build the `AiModel` into a `Provider`
  const gpt4o = yield* Gpt4o
  // Provide the implementation of `Completions` to `generateDadJoke`

>   yield* gpt4o.provide(generateDadJoke)

Reviewer: ```suggestion
  const res = yield* gpt4o.provide(generateDadJoke)
```

Author: Because `generateDadJoke` returns `void` in my sample program, I'm intentionally not binding these to a return variable.

})
```

Before we can finish satisfying requirements and make our program runnable, it is important that we understand the purpose of the `AiModel` data type.

## Understanding `AiModel`

The `AiModel` data type represents a **provider-specific implementation** of one or more services, such as `Completions` or `Embeddings`. It is the primary way that you can plug a real LLM (like OpenAI’s GPT-4o) into your program.

```ts showLineNumbers=false
export interface AiModel<Provides, Requires> {}
```

An `AiModel` has two generic type parameters:

- **Provides** - the services this model will provide when built
- **Requires** - the services this model will require to be built

This allows Effect to track which services should be added to the requirements of the program that builds the `AiModel`, as well as which services the built `AiModel` can provide.

### Creating an `AiModel`

To create an `AiModel`, you can use the model-specific factory from one of Effect's provider integration packages. 

> For example:

Reviewer: ```suggestion
**Example** (Defining an OpenAI-Powered `AiModel`)
```

```ts showLineNumbers=false
import { OpenAiCompletions } from "@effect/ai-openai"

//      ┌─── AiModel<Completions, OpenAiClient>
//      ▼
const Gpt4o = OpenAiCompletions.model("gpt-4o")
```

This creates an `AiModel` that:

- **Provides** an OpenAi-specific implementation of the `Completions` service using `"gpt-4o"`
- **Requires** an `OpenAiClient` to be built

### Building an `AiModel`

> When you build an `AiModel`, you get back a `Provider<Provides>`:

Reviewer: "build" sounds a bit off to me here, I'm just yielding, but "build" makes it sound like I'm doing something more involved

Author: The terminology is somewhat unfortunate I suppose, but that is precisely what we are doing - we're building the `AiModel` into a `Provider` which can provide the requisite services.

  //      ┌─── Provider<Completions>
  //      ▼
  const gpt4o = yield* Gpt4o
})
```

A `Provider` has a single `.provide(...)` method: 

```ts showLineNumbers=false
gpt4o.provide(myAiProgram)
```

The `.provide(...)` method injects the services built by the `AiModel` into an Effect program, thereby removing those services from that program's requirements.

### Benefits of `AiModel`

There are several benefits to this approach:

**Reusability**

You can call `.provide(...)` with the same built model as many times as you like.

For example, we can re-use the same built model to provide `Completions` to multiple calls to `generateDadJoke`:

```ts twoslash {16-18} collapse={5-10}
import { OpenAiCompletions } from "@effect/ai-openai"
import { Completions } from "@effect/ai"
import { Console, Effect } from "effect"

const generateDadJoke = Effect.gen(function*() {
  const completions = yield* Completions.Completions
  return yield* completions.create("Generate a hilarious dad joke").pipe(
    Effect.andThen((response) => Console.log(response.text))
  )
})

const Gpt4o = OpenAiCompletions.model("gpt-4o")

const main = Effect.gen(function*() {
  const gpt = yield* Gpt4o

>   yield* gpt.provide(generateDadJoke)
>   yield* gpt.provide(generateDadJoke)
>   yield* gpt.provide(generateDadJoke)

Reviewer: ```suggestion
  const res1 = yield* gpt.provide(generateDadJoke)
  const res2 = yield* gpt.provide(generateDadJoke)
  const res3 = yield* gpt.provide(generateDadJoke)
```

Author: Same comment as above.

})
```

**Flexibility**

If we know that one model or provider performs better at a given task than another, we can freely mix and match models and providers together. 

For example, if we know Anthropic's Claude generates some really great dad jokes, we can mix it into our existing program with just a few lines of code:

> 

Reviewer: ```suggestion
**Example** (Mixing Multiple Providers and Models)
```

> ```ts twoslash {1,14,20,23} collapse={6-11}

Reviewer: ```suggestion
```ts twoslash {23} collapse={6-11}
```

import { AnthropicCompletions } from "@effect/ai-anthropic"
import { OpenAiCompletions } from "@effect/ai-openai"
import { Completions } from "@effect/ai"
import { Console, Effect } from "effect"

const generateDadJoke = Effect.gen(function*() {
  const completions = yield* Completions.Completions
  return yield* completions.create("Generate a hilarious dad joke").pipe(
    Effect.andThen((response) => Console.log(response.text))
  )
})

const Gpt4o = OpenAiCompletions.model("gpt-4o")
const Claude37 = AnthropicCompletions.model("claude-3-7-sonnet-latest")

//      ┌─── Effect<void, AiError, AnthropicClient | OpenAiClient>
//      ▼
const main = Effect.gen(function*() {
  const gpt = yield* Gpt4o
  const claude = yield* Claude37
  yield* gpt.provide(generateDadJoke)
  yield* gpt.provide(generateDadJoke)
  yield* claude.provide(generateDadJoke)
})
```

> Because Effect performs type-level dpeendency tracking, we can see that an `AnthropicClient` must now also be provided to make our program runnable.

Reviewer: typo: "dpeendency"
